From a1124927929eb4b3ba3237017fc78aa413543968 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Tue, 24 Jul 2018 14:05:48 +0800
Subject: [PATCH 136/139] drm: sun4i: sun8i-hdmi-phy: add support for ddc-en
 gpio

Some boards have a gate GPIO attached to the level shifter transistor of
HDMI DDC pins.

Add support for controlling it.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h  |  2 ++
 drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c | 15 +++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
index 051666aa49b3..7c8ffbc54327 100644
--- a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
+++ b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
@@ -9,6 +9,7 @@
 #include <drm/bridge/dw_hdmi.h>
 #include <drm/drm_encoder.h>
 #include <linux/clk.h>
+#include <linux/of_gpio.h>
 #include <linux/regmap.h>
 #include <linux/regulator/consumer.h>
 #include <linux/reset.h>
@@ -172,6 +173,7 @@ struct sun8i_hdmi_phy {
 	unsigned int			rcal;
 	struct regmap			*regs;
 	struct reset_control		*rst_phy;
+	struct gpio_desc		*gpiod_ddc_en;
 	struct sun8i_hdmi_phy_variant	*variant;
 };
 
diff --git a/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c b/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
index 27f400119ac4..38d7091b603a 100644
--- a/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
+++ b/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
@@ -644,6 +644,18 @@ int sun8i_hdmi_phy_probe(struct sun8i_dw_hdmi *hdmi, struct device_node *node)
 		return PTR_ERR(phy->regs);
 	}
 
+	phy->gpiod_ddc_en = devm_gpiod_get_from_of_node(dev, node,
+						       "ddc-en-gpios", 0,
+						       GPIOD_OUT_LOW,
+						       "ddc-en");
+	if (IS_ERR(phy->gpiod_ddc_en)) {
+		dev_err(dev, "Couldn't request DDC enable GPIO\n");
+		return PTR_ERR(phy->gpiod_ddc_en);
+	}
+
+	if (phy->gpiod_ddc_en)
+		gpiod_set_value_cansleep(phy->gpiod_ddc_en, 1);
+
 	phy->clk_bus = of_clk_get_by_name(node, "bus");
 	if (IS_ERR(phy->clk_bus)) {
 		dev_err(dev, "Could not get bus clock\n");
@@ -741,6 +753,9 @@ void sun8i_hdmi_phy_remove(struct sun8i_dw_hdmi *hdmi)
 	clk_disable_unprepare(phy->clk_bus);
 	clk_disable_unprepare(phy->clk_phy);
 
+	if (phy->gpiod_ddc_en)
+		gpiod_set_value_cansleep(phy->gpiod_ddc_en, 0);
+
 	reset_control_assert(phy->rst_phy);
 
 	reset_control_put(phy->rst_phy);
-- 
2.18.0

